// @imgproxy/imgproxy-js-core/dist/imgproxy-js-core.mjs
const l = (t) => t === !0 || t === "t" || t === 1 ? "t" : "f", b = (t) => `${t}${t.includes(".") ? "" : " option"}`;
function n(t, e, o) {
  if (t === void 0)
    throw new Error(
      `${b(e)} is undefined${o ? `. ${o}` : ""}`
    );
}
function P(t, e, o, r) {
  return `${b(t)} value can't be less${r ? " or equal" : ""} then ${e}${o ? `. ${o}` : ""}`;
}
function i(t, e, o) {
  if (typeof t != "number")
    throw new Error(
      `${b(e)} is not a number${o != null && o.addInfo ? `. ${o == null ? void 0 : o.addInfo}` : ""}`
    );
  if (!(o != null && o.addParam))
    return;
  const r = o.addParam, s = o == null ? void 0 : o.addInfo;
  if (r.min !== void 0) {
    if (r.minEqual && t <= r.min)
      throw new Error(P(e, r.min, s, !0));
    if (t < r.min)
      throw new Error(P(e, r.min, s));
    if (r.max !== void 0 && t > r.max)
      throw new Error(
        `${b(e)} value can't be more than ${r.max}${s ? `. ${s}` : ""}`
      );
  }
  if (r.isInt && !Number.isInteger(t))
    throw new Error(`${b(e)} is must be an integer`);
}
function d(t, e, o) {
  const r = `Valid values are: ${Object.keys(t).join(", ")}`;
  if (!t[e])
    throw new Error(`${b(o)} is invalid. ${r}`);
}
function pe(t, e, o) {
  if (!t.includes(e)) {
    const r = `Valid values are: ${t.join(", ")}`;
    throw new Error(`${b(o)} is invalid. ${r}`);
  }
}
function _(t, e, o) {
  if (typeof t != "string")
    throw new Error(`${b(e)} is not a string`);
  if (o) {
    if (t.match(/[^0-9a-fA-F]/))
      throw new Error(`${b(e)} must be hexadecimal`);
    if (t.length !== 3 && t.length !== 6 && t.length !== 8)
      throw new Error(
        `${b(e)} must be 3, 6 or 8 characters long (with alpha)`
      );
  }
}
function $(t, e) {
  if (!Array.isArray(t))
    throw new Error(`${b(e)} is not an array`);
  if (t.length === 0)
    throw new Error(`${b(e)} is empty`);
}
function m(t, e) {
  if (typeof t != "boolean")
    throw new Error(`${b(e)} is not a boolean`);
}
const k = (t) => t.adjust || t.a, me = (t) => !!k(t), ge = (t) => {
  const e = k(t);
  n(e, "adjust"), e.brightness !== void 0 && i(e.brightness, "adjust.brightness", {
    addParam: { min: -255, max: 255 }
  });
  const o = e.brightness ?? "", r = e.contrast ?? "", s = e.saturation ?? "";
  return `a:${o}:${r}:${s}`;
}, fe = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: ge,
  test: me
}, Symbol.toStringTag, { value: "Module" })), $e = {
  none: !0,
  size: !0,
  dssim: !0,
  ml: !0
}, B = (t) => t.autoquality || t.aq, ye = (t) => !!B(t), Oe = (t) => {
  const e = B(t);
  n(e, "autoquality");
  const { method: o, target: r, min_quality: s, max_quality: a, allowed_error: c } = e;
  if (o && d($e, o, "autoquality.method"), r !== void 0 && i(r, "autoquality.target", {
    addParam: { min: 0 }
  }), s !== void 0 && i(s, "autoquality.min_quality", {
    addParam: { min: 0, max: 100 }
  }), a !== void 0 && i(a, "autoquality.max_quality", {
    addParam: { min: 0, max: 100 }
  }), c !== void 0) {
    if (o !== "dssim" && o !== "ml")
      throw new Error(
        "autoquality.allowed_error is applicable only to dssim and ml methods"
      );
    i(c, "autoquality.allowed_error", {
      addParam: { min: 0, max: 1 }
    });
  }
  return `aq:${o ?? ""}:${r ?? ""}:${s ?? ""}:${a ?? ""}:${c ?? ""}`.replace(/:+$/, "");
}, ve = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Oe,
  test: ye
}, Symbol.toStringTag, { value: "Module" })), q = (t) => {
  if ("auto_rotate" in t)
    return t.auto_rotate;
  if ("ar" in t)
    return t.ar;
}, he = (t) => q(t) !== void 0, je = (t) => {
  const e = q(t);
  return n(e, "auto_rotate"), `ar:${l(e)}`;
}, Se = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: je,
  test: he
}, Symbol.toStringTag, { value: "Module" })), E = (t) => t.background || t.bg, we = (t) => !!E(t), Pe = (t) => {
  const e = E(t);
  if (n(e, "background"), typeof e == "number")
    throw new Error("background option is not a string or object");
  return typeof e == "string" ? (_(e, "background", !0), `bg:${e}`) : (n(e.r, "background.r"), n(e.g, "background.g"), n(e.b, "background.b"), i(e.r, "background.r"), i(e.g, "background.g"), i(e.b, "background.b"), `bg:${e.r}:${e.g}:${e.b}`);
}, ze = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Pe,
  test: we
}, Symbol.toStringTag, { value: "Module" })), I = (t) => t.background_alpha ?? t.bga, xe = (t) => !!I(t), Te = (t) => {
  const e = I(t);
  return n(e, "background_alpha"), i(e, "background_alpha", {
    addParam: { min: 0, max: 1 }
  }), `background_alpha:${e}`;
}, Me = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Te,
  test: xe
}, Symbol.toStringTag, { value: "Module" })), A = (t) => t.blur ?? t.bl, ke = (t) => !!A(t), Be = (t) => {
  const e = A(t);
  return n(e, "blur"), i(e, "blur", { addParam: { min: 0 } }), `bl:${e}`;
}, qe = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Be,
  test: ke
}, Symbol.toStringTag, { value: "Module" })), V = (t) => t.blur_detections || t.bd, Ee = (t) => !!V(t), Ie = (t) => {
  const e = V(t);
  n(e, "blur_detections"), n(e.sigma, "blur_detections.sigma"), i(e.sigma, "blur_detections.sigma");
  const o = e.sigma, r = e.class_names ? e.class_names.join(":") : "";
  return `bd:${o}:${r}`.replace(/:+$/, "");
}, Ae = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Ie,
  test: Ee
}, Symbol.toStringTag, { value: "Module" })), F = (t) => t.brightness ?? t.br, Ve = (t) => !!F(t), Fe = (t) => {
  const e = F(t);
  return n(e, "brightness"), i(e, "brightness", {
    addParam: { min: -255, max: 255 }
  }), `br:${e}`;
}, Ue = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Fe,
  test: Ve
}, Symbol.toStringTag, { value: "Module" })), U = (t) => t.cachebuster || t.cb, Ne = (t) => !!U(t), Re = (t) => {
  const e = U(t);
  return n(e, "cachebuster"), _(e, "cachebuster"), `cb:${e}`;
}, N = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Re,
  test: Ne
}, Symbol.toStringTag, { value: "Module" })), R = (t) => t.contrast ?? t.co, Ce = (t) => !!R(t), Qe = (t) => {
  const e = R(t);
  return n(e, "contrast"), i(e, "contrast", { addParam: { min: 0 } }), `co:${e}`;
}, De = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Qe,
  test: Ce
}, Symbol.toStringTag, { value: "Module" })), C = {
  no: !0,
  so: !0,
  ea: !0,
  we: !0,
  ce: !0,
  noea: !0,
  nowe: !0,
  soea: !0,
  sowe: !0
}, He = {
  ...C,
  sm: !0,
  fp: !0,
  obj: !0
}, Q = (t) => t.gravity || t.g, O = (t) => !!Q(t), y = (t, e) => `${e ? "" : "g:"}${t.replace(/:+$/, "")}`, v = (t, e = {}) => {
  const { headless: o = !1 } = e, r = Q(t);
  n(r, "gravity"), n(r.type, "gravity.type");
  const s = r.type;
  if (d(He, s, "gravity.type"), (r.x_offset || r.y_offset) && d(C, s, "gravity.type"), r.class_names && s !== "obj")
    throw new Error("gravity.class_names can be used only with type obj");
  if ((r.x || r.y) && s !== "fp")
    throw new Error("gravity.x and gravity.y can be used only with type fp");
  if (s === "sm")
    return y(s, o);
  if (s === "fp") {
    const a = r;
    return n(a.x, "gravity.x"), n(a.y, "gravity.y"), i(a.x, "gravity.x", { addParam: { min: 0, max: 1 } }), i(a.y, "gravity.y", { addParam: { min: 0, max: 1 } }), y(`${s}:${a.x}:${a.y}`, o);
  }
  if (s === "obj") {
    const a = r;
    n(r.class_names, "gravity.class_names"), $(a.class_names, "gravity.class_names");
    const c = a.class_names;
    return y(`${s}:${c.join(":")}`, o);
  } else {
    const a = r, c = a.x_offset === void 0 ? "" : a.x_offset, u = a.y_offset === void 0 ? "" : a.y_offset;
    return y(`${a.type}:${c}:${u}`, o);
  }
}, Xe = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: v,
  test: O
}, Symbol.toStringTag, { value: "Module" })), D = (t) => t.crop || t.c, We = (t) => !!D(t), Je = (t) => {
  const e = D(t);
  n(e, "crop");
  const { width: o, height: r } = e;
  n(o, "crop.width"), n(r, "crop.height"), i(o, "crop.width", { addParam: { min: 0 } }), i(r, "crop.height", { addParam: { min: 0 } });
  const s = O(e) ? v(e, { headless: !0 }) : "";
  return `c:${o}:${r}:${s}`.replace(/:+$/, "");
}, H = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Je,
  test: We
}, Symbol.toStringTag, { value: "Module" })), X = (t) => {
  if ("disable_animation" in t)
    return t.disable_animation;
  if ("da" in t)
    return t.da;
}, Ye = (t) => X(t) !== void 0, Ge = (t) => {
  const e = X(t);
  return n(e, "disable_animation"), `da:${l(e)}`;
}, Ke = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Ge,
  test: Ye
}, Symbol.toStringTag, { value: "Module" })), W = (t) => t.dpi, Le = (t) => !!W(t), Ze = (t) => {
  const e = W(t);
  return n(e, "dpi"), i(e, "dpi"), `dpi:${e}`;
}, to = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Ze,
  test: Le
}, Symbol.toStringTag, { value: "Module" })), J = (t) => t.dpr, eo = (t) => !!J(t), oo = (t) => {
  const e = J(t);
  return n(e, "dpr"), i(e, "dpr", { addParam: { min: 0 } }), `dpr:${e}`;
}, ro = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: oo,
  test: eo
}, Symbol.toStringTag, { value: "Module" })), Y = (t) => t.draw_detections || t.dd, no = (t) => !!Y(t), so = (t) => {
  const e = Y(t);
  n(e, "draw_detections"), n(e.draw, "draw_detections.draw");
  const o = l(e.draw), r = e.class_names ? `:${e.class_names.join(":")}` : "";
  return `dd:${o}${r}`;
}, io = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: so,
  test: no
}, Symbol.toStringTag, { value: "Module" })), G = (t) => {
  if ("enforce_thumbnail" in t)
    return t.enforce_thumbnail;
  if ("eth" in t)
    return t.eth;
}, ao = (t) => G(t) !== void 0, lo = (t) => {
  const e = G(t);
  return n(e, "enforce_thumbnail"), `eth:${l(e)}`;
}, co = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: lo,
  test: ao
}, Symbol.toStringTag, { value: "Module" })), K = (t) => {
  if ("enlarge" in t)
    return t.enlarge;
  if ("el" in t)
    return t.el;
}, uo = (t) => K(t) !== void 0, bo = (t) => {
  const e = K(t);
  return n(e, "enlarge"), `el:${l(e)}`;
}, _o = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: bo,
  test: uo
}, Symbol.toStringTag, { value: "Module" })), L = (t) => t.expires || t.exp, po = (t) => !!L(t), mo = (t) => {
  const e = L(t);
  return n(e, "expires"), i(e, "expires"), `exp:${e}`;
}, Z = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: mo,
  test: po
}, Symbol.toStringTag, { value: "Module" })), tt = (t) => t.extend || t.ex, j = (t) => !!tt(t), go = (t, e) => `${e ? "" : "ex:"}${t.replace(/:+$/, "")}`, S = (t, e = {}) => {
  const { headless: o = !1 } = e, r = tt(t);
  n(r, "extend"), n(r.extend, "extend.extend");
  const s = O(r) ? `:${v(r, { headless: !0 })}` : "";
  return go(`${l(r.extend)}${s}`, o);
}, fo = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: S,
  test: j
}, Symbol.toStringTag, { value: "Module" })), et = (t) => t.extend_aspect_ratio || t.exar || t.extend_ar, $o = (t) => !!et(t), yo = (t) => {
  const e = et(t);
  n(e, "extend_aspect_ratio"), n(e.extend, "extend_aspect_ratio.extend");
  const o = O(e) ? `:${v(e, { headless: !0 })}` : "";
  return `exar:${l(e.extend)}${o}`;
}, Oo = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: yo,
  test: $o
}, Symbol.toStringTag, { value: "Module" })), ot = (t) => t.fallback_image_url || t.fiu, vo = (t) => !!ot(t), ho = (t) => {
  const e = ot(t);
  return n(e, "fallback_image_url"), _(e, "fallback_image_url"), `fiu:${e}`;
}, jo = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: ho,
  test: vo
}, Symbol.toStringTag, { value: "Module" })), rt = (t) => t.filename || t.fn, So = (t) => !!rt(t), wo = (t) => {
  const e = rt(t);
  n(e, "filename"), n(e.name, "filename.name"), _(e.name, "filename.name");
  const o = e.encoded === void 0 ? "" : `:${l(e.encoded)}`;
  return `fn:${e.name}${o}`;
}, Po = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: wo,
  test: So
}, Symbol.toStringTag, { value: "Module" })), zo = {
  png: !0,
  jpg: !0,
  webp: !0,
  avif: !0,
  gif: !0,
  ico: !0,
  svg: !0,
  bmp: !0,
  tiff: !0,
  mp4: !0,
  best: !0
}, nt = (t) => t.format || t.f || t.ext, xo = (t) => !!nt(t), To = (t) => {
  const e = nt(t);
  return n(e, "format"), d(zo, e, "format"), `f:${e}`;
}, Mo = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: To,
  test: xo
}, Symbol.toStringTag, { value: "Module" })), st = (t) => t.format_quality || t.fq, ko = (t) => !!st(t), Bo = (t) => {
  const e = st(t);
  n(e, "format_quality"), $(e, "format_quality");
  const o = [];
  return e.forEach((r) => {
    o.push(`${r.format}:${r.quality}`);
  }), `fq:${o.join(":")}`;
}, qo = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Bo,
  test: ko
}, Symbol.toStringTag, { value: "Module" })), Eo = {
  down: !0,
  up: !0,
  right: !0,
  left: !0
}, it = (t) => t.gradient || t.gr, Io = (t) => !!it(t), Ao = (t) => {
  const e = it(t);
  n(e, "gradient");
  const { opacity: o, color: r, direction: s, start: a, stop: c } = e;
  return n(o, "gradient.opacity"), i(o, "gradient.opacity", {
    addParam: { min: 0, max: 1 }
  }), r && _(r, "gradient.color", !0), s && (_(s, "gradient.direction"), d(Eo, s, "gradient.direction")), a && i(a, "gradient.start", { addParam: { min: 0, max: 1 } }), c && i(c, "gradient.stop", { addParam: { min: 0, max: 1 } }), `gr:${o}:${r || ""}:${s || ""}:${a || ""}:${c || ""}`.replace(/:+$/, "");
}, Vo = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Ao,
  test: Io
}, Symbol.toStringTag, { value: "Module" })), at = (t) => t.height ?? t.h, Fo = (t) => !!at(t), Uo = (t) => {
  const e = at(t);
  return n(e, "height"), i(e, "height", { addParam: { min: 0 } }), `h:${e}`;
}, No = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Uo,
  test: Fo
}, Symbol.toStringTag, { value: "Module" })), lt = (t) => t.jpeg_options || t.jpgo, Ro = (t) => !!lt(t), Co = (t) => {
  const e = lt(t);
  n(e, "jpeg_options");
  const {
    progressive: o,
    no_subsample: r,
    trellis_quant: s,
    overshoot_deringing: a,
    optimize_scans: c,
    quant_table: u
  } = e;
  if (o && m(o, "jpeg_options.progressive"), r && m(r, "jpeg_options.no_subsample"), s && m(s, "jpeg_options.trellis_quant"), a && m(a, "jpeg_options.overshoot_deringing"), c && (m(c, "jpeg_options.optimize_scans"), o === !1))
    throw new Error(
      "jpeg_options.progressive must be true if jpeg_options.optimize_scans is true"
    );
  return u && i(u, "jpeg_options.quant_table", {
    addParam: { min: 0, max: 8 }
  }), `jpgo:${o === void 0 ? "" : o}:${r === void 0 ? "" : r}:${s === void 0 ? "" : s}:${a === void 0 ? "" : a}:${c === void 0 ? "" : c}:${u === void 0 ? "" : u}`.replace(/:+$/, "");
}, Qo = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Co,
  test: Ro
}, Symbol.toStringTag, { value: "Module" })), ct = (t) => {
  if ("keep_copyright" in t)
    return t.keep_copyright;
  if ("kcr" in t)
    return t.kcr;
}, Do = (t) => ct(t) !== void 0, Ho = (t) => {
  const e = ct(t);
  return n(e, "keep_copyright"), `kcr:${l(e)}`;
}, Xo = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Ho,
  test: Do
}, Symbol.toStringTag, { value: "Module" })), ut = (t) => {
  if ("max_animation_frames" in t)
    return t.max_animation_frames;
  if ("maf" in t)
    return t.maf;
}, Wo = (t) => ut(t) !== void 0, Jo = (t) => {
  const e = ut(t);
  return n(e, "max_animation_frames"), i(e, "max_animation_frames", {
    addParam: { min: 0, minEqual: !0 }
  }), `maf:${e}`;
}, Yo = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Jo,
  test: Wo
}, Symbol.toStringTag, { value: "Module" })), dt = (t) => {
  if ("max_animation_frame_resolution" in t)
    return t.max_animation_frame_resolution;
  if ("mafr" in t)
    return t.mafr;
}, Go = (t) => dt(t) !== void 0, Ko = (t) => {
  const e = dt(t);
  return n(e, "max_animation_frame_resolution"), i(e, "max_animation_frame_resolution", {
    addParam: { min: 0 }
  }), `mafr:${e}`;
}, Lo = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Ko,
  test: Go
}, Symbol.toStringTag, { value: "Module" })), bt = (t) => t.max_bytes ?? t.mb, Zo = (t) => !!bt(t), tr = (t) => {
  const e = bt(t);
  return n(e, "max_bytes"), i(e, "max_bytes", { addParam: { min: 0 } }), `mb:${e}`;
}, er = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: tr,
  test: Zo
}, Symbol.toStringTag, { value: "Module" })), _t = [
  "no",
  "so",
  "ea",
  "we",
  "noea",
  "nowe",
  "soea",
  "sowe",
  "ce"
], or = (t) => _t.includes(t.type), pt = (t) => t.objects_position || t.obj_pos || t.op, rr = (t) => !!pt(t), nr = (t) => {
  const e = pt(t);
  n(e, "objects_position");
  const o = e.type;
  if (pe([..._t, "fp", "prop"], o, "type"), or(e)) {
    e.xOffset !== void 0 && i(e.xOffset, "objects_position.xOffset"), e.yOffset !== void 0 && i(e.yOffset, "objects_position.yOffset");
    const r = e.xOffset ?? "", s = e.yOffset ?? "";
    return `op:${o}:${r}:${s}`;
  } else if (e.type === "fp") {
    e.x !== void 0 && i(e.x, "objects_position.x", {
      addParam: { min: 0, max: 1 }
    }), e.y !== void 0 && i(e.y, "objects_position.y", {
      addParam: { min: 0, max: 1 }
    });
    const r = e.x ?? "", s = e.y ?? "";
    return `op:${o}:${r}:${s}`;
  } else {
    if (o === "prop")
      return "op:prop";
    throw new Error(`Unknown object position type: ${o}`);
  }
}, sr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: nr,
  test: rr
}, Symbol.toStringTag, { value: "Module" })), mt = (t) => {
  if ("max_src_file_size" in t)
    return t.max_src_file_size;
  if ("msfs" in t)
    return t.msfs;
}, ir = (t) => mt(t) !== void 0, ar = (t) => {
  const e = mt(t);
  return n(e, "max_src_file_size"), i(e, "max_src_file_size", { addParam: { min: 0 } }), `msfs:${e}`;
}, gt = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: ar,
  test: ir
}, Symbol.toStringTag, { value: "Module" })), ft = (t) => {
  if ("max_src_resolution" in t)
    return t.max_src_resolution;
  if ("msr" in t)
    return t.msr;
}, lr = (t) => ft(t) !== void 0, cr = (t) => {
  const e = ft(t);
  return n(e, "max_src_resolution"), i(e, "max_src_resolution", {
    addParam: { min: 0, minEqual: !0 }
  }), `msr:${e}`;
}, $t = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: cr,
  test: lr
}, Symbol.toStringTag, { value: "Module" })), yt = (t) => t.min_height ?? t.mh, ur = (t) => !!yt(t), dr = (t) => {
  const e = yt(t);
  return n(e, "min_height"), i(e, "min_height", { addParam: { min: 0 } }), `mh:${e}`;
}, br = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: dr,
  test: ur
}, Symbol.toStringTag, { value: "Module" })), Ot = (t) => t.min_width ?? t.mw, _r = (t) => !!Ot(t), pr = (t) => {
  const e = Ot(t);
  return n(e, "min_width"), i(e, "min_width", { addParam: { min: 0 } }), `mw:${e}`;
}, mr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: pr,
  test: _r
}, Symbol.toStringTag, { value: "Module" })), vt = (t) => t.padding || t.pd, gr = (t) => !!vt(t), fr = (t) => {
  const e = vt(t);
  if (n(e, "padding"), typeof e == "string")
    throw new Error("padding option is not a number or object");
  if (typeof e == "number")
    return i(e, "padding", { addParam: { min: 0 } }), `pd:${e}`;
  e.top !== void 0 && i(e.top, "padding.top"), e.right !== void 0 && i(e.right, "padding.right"), e.bottom !== void 0 && i(e.bottom, "padding.bottom"), e.left !== void 0 && i(e.left, "padding.left");
  const o = e.top ?? "", r = e.right ?? "", s = e.bottom ?? "", a = e.left ?? "";
  return `pd:${o}:${r}:${s}:${a}`.replace(/:+$/, "");
}, $r = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: fr,
  test: gr
}, Symbol.toStringTag, { value: "Module" })), ht = (t) => {
  if ("page" in t)
    return t.page;
  if ("pg" in t)
    return t.pg;
}, yr = (t) => ht(t) !== void 0, Or = (t) => {
  const e = ht(t);
  return n(e, "page"), i(e, "page", { addParam: { min: 0, isInt: !0 } }), `pg:${e}`;
}, jt = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Or,
  test: yr
}, Symbol.toStringTag, { value: "Module" })), St = (t) => {
  if ("pages" in t)
    return t.pages;
  if ("pgs" in t)
    return t.pgs;
}, vr = (t) => St(t) !== void 0, hr = (t) => {
  const e = St(t);
  return n(e, "pages"), i(e, "pages", { addParam: { min: 1, isInt: !0 } }), `pgs:${e}`;
}, jr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: hr,
  test: vr
}, Symbol.toStringTag, { value: "Module" })), wt = (t) => t.pixelate ?? t.pix, Sr = (t) => !!wt(t), wr = (t) => {
  const e = wt(t);
  return n(e, "pixelate"), i(e, "pixelate", { addParam: { min: 0 } }), `pix:${e}`;
}, Pr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: wr,
  test: Sr
}, Symbol.toStringTag, { value: "Module" })), Pt = (t) => t.png_options || t.pngo, zr = (t) => !!Pt(t), xr = (t) => {
  const e = Pt(t);
  n(e, "png_options");
  const { interlaced: o, quantize: r, quantization_colors: s } = e;
  return o && m(o, "png_options.interlaced"), r && m(r, "png_options.quantize"), s && i(s, "png_options.quantization_colors", {
    addParam: { min: 2, max: 256 }
  }), `pngo:${o === void 0 ? "" : o}:${r === void 0 ? "" : r}:${s || ""}`.replace(/:+$/, "");
}, Tr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: xr,
  test: zr
}, Symbol.toStringTag, { value: "Module" })), zt = (t) => t.preset || t.pr, Mr = (t) => !!zt(t), kr = (t) => {
  const e = zt(t);
  if (n(e, "preset"), $(e, "preset"), e.some((o) => typeof o != "string"))
    throw new Error("preset option should contain only strings");
  return `pr:${e.join(":")}`;
}, xt = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: kr,
  test: Mr
}, Symbol.toStringTag, { value: "Module" })), Tt = (t) => t.quality ?? t.q, Br = (t) => !!Tt(t), qr = (t) => {
  const e = Tt(t);
  return n(e, "quality"), i(e, "quality", { addParam: { min: 0, max: 100 } }), `q:${e}`;
}, Er = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: qr,
  test: Br
}, Symbol.toStringTag, { value: "Module" })), Mt = (t) => {
  if ("raw" in t)
    return t.raw;
}, Ir = (t) => Mt(t) !== void 0, Ar = (t) => {
  const e = Mt(t);
  return n(e, "raw"), `raw:${l(e)}`;
}, Vr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Ar,
  test: Ir
}, Symbol.toStringTag, { value: "Module" })), Fr = {
  fit: !0,
  fill: !0,
  auto: !0,
  "fill-down": !0,
  force: !0
}, kt = (t) => t.resize || t.rs, Ur = (t) => !!kt(t), Nr = (t) => {
  const e = kt(t);
  n(e, "resize");
  const { resizing_type: o, width: r, height: s, enlarge: a } = e;
  o && d(Fr, o, "resize.resizing_type"), r && i(r, "resize.width", { addParam: { min: 0 } }), s && i(s, "resize.height", { addParam: { min: 0 } });
  const c = o || "", u = r || "", p = s || "", g = a === void 0 ? "" : l(a), f = j(e) ? S(e, { headless: !0 }) : "";
  return `rs:${c}:${u}:${p}:${g}:${f}`.replace(/:+$/, "");
}, Rr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Nr,
  test: Ur
}, Symbol.toStringTag, { value: "Module" })), Cr = {
  nearest: !0,
  linear: !0,
  cubic: !0,
  lanczos2: !0,
  lanczos3: !0
}, Bt = (t) => t.resizing_algorithm || t.ra, Qr = (t) => !!Bt(t), Dr = (t) => {
  const e = Bt(t);
  return n(e, "resizing_algorithm"), d(Cr, e, "resizing_algorithm"), `ra:${e}`;
}, Hr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Dr,
  test: Qr
}, Symbol.toStringTag, { value: "Module" })), Xr = {
  fit: !0,
  fill: !0,
  "fill-down": !0,
  force: !0,
  auto: !0
}, qt = (t) => t.resizing_type || t.rt, Wr = (t) => !!qt(t), Jr = (t) => {
  const e = qt(t);
  return n(e, "resizing_type"), d(Xr, e, "resizing_type"), `rt:${e}`;
}, Yr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Jr,
  test: Wr
}, Symbol.toStringTag, { value: "Module" })), Et = (t) => {
  if ("return_attachment" in t)
    return t.return_attachment;
  if ("att" in t)
    return t.att;
}, Gr = (t) => Et(t) !== void 0, Kr = (t) => {
  const e = Et(t);
  return n(e, "return_attachment"), `att:${l(e)}`;
}, Lr = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Kr,
  test: Gr
}, Symbol.toStringTag, { value: "Module" })), z = {
  0: !0,
  90: !0,
  180: !0,
  270: !0
}, It = (t) => t.rotate ?? t.rot, Zr = (t) => !!It(t), tn = (t) => {
  const e = It(t), o = `You can use values ${Object.keys(z).join(", ")}`;
  return n(e, "rotate"), i(e, "rotate", { addInfo: o }), d(z, e, "rotate"), `rot:${e}`;
}, en = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: tn,
  test: Zr
}, Symbol.toStringTag, { value: "Module" })), At = (t) => t.saturation ?? t.sa, on = (t) => !!At(t), rn = (t) => {
  const e = At(t);
  return n(e, "saturation"), i(e, "saturation", { addParam: { min: 0 } }), `sa:${e}`;
}, nn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: rn,
  test: on
}, Symbol.toStringTag, { value: "Module" })), Vt = (t) => t.sharpen ?? t.sh, sn = (t) => !!Vt(t), an = (t) => {
  const e = Vt(t);
  return n(e, "sharpen"), i(e, "sharpen", { addParam: { min: 0 } }), `sh:${e}`;
}, ln = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: an,
  test: sn
}, Symbol.toStringTag, { value: "Module" })), Ft = (t) => t.size || t.s, cn = (t) => !!Ft(t), un = (t) => {
  const e = Ft(t);
  n(e, "size");
  const { width: o, height: r, enlarge: s } = e;
  o && i(o, "size.width", { addParam: { min: 0 } }), r && i(r, "size.height", { addParam: { min: 0 } });
  const a = o || "", c = r || "", u = s === void 0 ? "" : l(s), p = j(e) ? S(e, { headless: !0 }) : "";
  return `s:${a}:${c}:${u}:${p}`.replace(/:+$/, "");
}, dn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: un,
  test: cn
}, Symbol.toStringTag, { value: "Module" })), x = [
  "jpg",
  "png",
  "webp",
  "avif",
  "gif",
  "ico",
  "svg",
  "heic",
  "bmp",
  "tiff",
  "pdf",
  "mp4"
], Ut = (t) => t.skip_processing || t.sp, bn = (t) => !!Ut(t), _n = (t) => {
  const e = Ut(t);
  if (n(e, "skip_processing"), $(e, "skip_processing"), e.some((o) => !x.includes(o))) {
    const o = `Valid values are: ${x.join(",")}`;
    throw new Error(
      `skip_processing option contains unsupported extension. ${o}`
    );
  }
  return `sp:${e.join(":")}`;
}, pn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: _n,
  test: bn
}, Symbol.toStringTag, { value: "Module" })), Nt = (t) => {
  if ("strip_color_profile" in t)
    return t.strip_color_profile;
  if ("scp" in t)
    return t.scp;
}, mn = (t) => Nt(t) !== void 0, gn = (t) => {
  const e = Nt(t);
  return n(e, "strip_color_profile"), `scp:${l(e)}`;
}, fn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: gn,
  test: mn
}, Symbol.toStringTag, { value: "Module" })), Rt = (t) => "strip_metadata" in t ? t.strip_metadata : "sm" in t ? t.sm : void 0, $n = (t) => Rt(t) !== void 0, yn = (t) => {
  const e = Rt(t);
  return n(e, "strip_metadata"), `sm:${l(e)}`;
}, On = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: yn,
  test: $n
}, Symbol.toStringTag, { value: "Module" })), Ct = (t) => t.style || t.stl, vn = (t) => !!Ct(t), hn = (t) => {
  const e = Ct(t);
  return n(e, "style"), _(e, "style"), `stl:${e}`;
}, jn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: hn,
  test: vn
}, Symbol.toStringTag, { value: "Module" })), Qt = (t) => t.trim || t.t, Sn = (t) => !!Qt(t), wn = (t) => {
  const e = Qt(t);
  n(e, "trim");
  const { threshold: o, color: r, equal_hor: s, equal_ver: a } = e;
  n(o, "trim.threshold"), i(o, "trim.threshold"), r && _(r, "trim.color", !0);
  const c = s === void 0 ? "" : l(s), u = a === void 0 ? "" : l(a);
  return `t:${o}:${r || ""}:${c}:${u}`.replace(/:+$/, "");
}, Pn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: wn,
  test: Sn
}, Symbol.toStringTag, { value: "Module" })), zn = {
  auto: !0,
  none: !0,
  always: !0
}, Dt = (t) => t.unsharp_masking || t.ush, xn = (t) => !!Dt(t), Tn = (t) => {
  const e = Dt(t);
  n(e, "unsharp_masking");
  const { mode: o, weight: r, divider: s } = e;
  return o && d(zn, o, "unsharp_masking.mode"), r !== void 0 && i(r, "unsharp_masking.weight", {
    addParam: { min: 0, minEqual: !0 }
  }), s !== void 0 && i(s, "unsharp_masking.divider", {
    addParam: { min: 0, minEqual: !0 }
  }), `ush:${o || ""}:${r || ""}:${s || ""}`.replace(/:+$/, "");
}, Mn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Tn,
  test: xn
}, Symbol.toStringTag, { value: "Module" })), Ht = (t) => {
  if ("video_thumbnail_second" in t)
    return t.video_thumbnail_second;
  if ("vts" in t)
    return t.vts;
}, kn = (t) => Ht(t) !== void 0, Bn = (t) => {
  const e = Ht(t);
  return n(e, "video_thumbnail_second"), i(e, "video_thumbnail_second", {
    addParam: { min: 1, isInt: !0 }
  }), `vts:${e}`;
}, Xt = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Bn,
  test: kn
}, Symbol.toStringTag, { value: "Module" })), qn = {
  ce: !0,
  no: !0,
  so: !0,
  ea: !0,
  we: !0,
  noea: !0,
  nowe: !0,
  soea: !0,
  sowe: !0,
  re: !0
}, Wt = (t) => t.watermark || t.wm, En = (t) => !!Wt(t), In = (t) => {
  const e = Wt(t);
  n(e, "watermark");
  const { opacity: o, position: r, x_offset: s, y_offset: a, scale: c } = e;
  return n(o, "watermark.opacity"), i(o, "watermark.opacity", { addParam: { min: 0, max: 1 } }), r && (_(r, "watermark.position"), d(qn, r, "watermark.position")), s && i(s, "watermark.x_offset"), a && i(a, "watermark.y_offset"), c && i(c, "watermark.scale"), `wm:${o}:${r || ""}:${s || ""}:${a || ""}:${c || ""}`.replace(/:+$/, "");
}, An = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: In,
  test: En
}, Symbol.toStringTag, { value: "Module" })), Jt = (t) => t.watermark_shadow ?? t.wmsh, Vn = (t) => !!Jt(t), Fn = (t) => {
  const e = Jt(t);
  return n(e, "watermark_shadow"), i(e, "watermark_shadow", {
    addParam: { min: 0 }
  }), `wmsh:${e}`;
}, Un = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Fn,
  test: Vn
}, Symbol.toStringTag, { value: "Module" })), Yt = (t) => t.watermark_size || t.wms, Nn = (t) => !!Yt(t), Rn = (t) => {
  const e = Yt(t);
  n(e, "watermark_size");
  const { width: o, height: r } = e;
  return o && i(o, "watermark_size.width", { addParam: { min: 0 } }), r && i(r, "watermark_size.height", { addParam: { min: 0 } }), `wms:${o || ""}:${r || ""}`.replace(/:+$/, "");
}, Cn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Rn,
  test: Nn
}, Symbol.toStringTag, { value: "Module" })), Gt = (t) => t.watermark_text || t.wmt, Qn = (t) => !!Gt(t), Dn = (t) => {
  const e = Gt(t);
  return n(e, "watermark_text"), _(e, "watermark_text"), `wmt:${e}`;
}, Hn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Dn,
  test: Qn
}, Symbol.toStringTag, { value: "Module" })), Kt = (t) => t.watermark_url || t.wmu, Xn = (t) => !!Kt(t), Wn = (t) => {
  const e = Kt(t);
  return n(e, "watermark_url"), _(e, "watermark_url"), `wmu:${e}`;
}, Jn = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Wn,
  test: Xn
}, Symbol.toStringTag, { value: "Module" })), Yn = {
  lossy: !0,
  near_lossless: !0,
  lossless: !0
}, Lt = (t) => t.webp_options || t.webpo, Gn = (t) => !!Lt(t), Kn = (t) => {
  const e = Lt(t);
  return n(e, "webp_options"), d(Yn, e, "webp_options"), `webpo:${e}`;
}, Ln = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Kn,
  test: Gn
}, Symbol.toStringTag, { value: "Module" })), Zt = (t) => t.width ?? t.w, Zn = (t) => !!Zt(t), ts = (t) => {
  const e = Zt(t);
  return n(e, "width"), i(e, "width", { addParam: { min: 0 } }), `w:${e}`;
}, es = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: ts,
  test: Zn
}, Symbol.toStringTag, { value: "Module" })), h = (t, e) => {
  n(t, e), i(t, e, { addParam: { min: 0 } });
}, te = (t) => t.zoom || t.z, os = (t) => !!te(t), rs = (t) => {
  const e = te(t);
  return n(e, "zoom"), typeof e == "string" && i(e, "zoom"), typeof e == "number" ? (h(e, "zoom"), `z:${e}`) : (h(e.zoom_x, "zoom.zoom_x"), h(e.zoom_y, "zoom.zoom_y"), `z:${e.zoom_x}:${e.zoom_y}`);
}, ns = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: rs,
  test: os
}, Symbol.toStringTag, { value: "Module" })), ss = (t) => t.wild_options, is = (t) => !!ss(t), as = (t) => {
  const e = t.wild_options;
  if (n(e, "raw_options"), $(e, "raw_options"), !e.every((r) => Array.isArray(r)))
    throw new Error("raw_options is not an array of arrays");
  let o = "";
  for (let r = 0; r < e.length; r++) {
    let s = "";
    for (let a = 0; a < e[r].length; a++)
      a > 0 && (s += ":"), s += e[r][a];
    r > 0 && (o += "/"), o += s;
  }
  return o;
}, ee = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: as,
  test: is
}, Symbol.toStringTag, { value: "Module" })), ls = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  adjust: fe,
  autoRotate: Se,
  autoquality: ve,
  background: ze,
  backgroundAlpha: Me,
  blur: qe,
  blurDetections: Ae,
  brightness: Ue,
  cacheBuster: N,
  contrast: De,
  crop: H,
  disableAnimation: Ke,
  dpi: to,
  dpr: ro,
  drawDetections: io,
  enforceThumbnail: co,
  enlarge: _o,
  expires: Z,
  extend: fo,
  extendAspectRatio: Oo,
  fallbackImageUrl: jo,
  filename: Po,
  format: Mo,
  formatQuality: qo,
  gradient: Vo,
  gravity: Xe,
  height: No,
  jpegOptions: Qo,
  keepCopyright: Xo,
  maxAnimationFrameResolution: Lo,
  maxAnimationFrames: Yo,
  maxBytes: er,
  maxSrcFileSize: gt,
  maxSrcResolution: $t,
  minHeight: br,
  minWidth: mr,
  objectPosition: sr,
  padding: $r,
  page: jt,
  pages: jr,
  pixelate: Pr,
  pngOptions: Tr,
  preset: xt,
  quality: Er,
  raw: Vr,
  resize: Rr,
  resizingAlgorithm: Hr,
  resizingType: Yr,
  returnAttachment: Lr,
  rotate: en,
  saturation: nn,
  sharpen: ln,
  size: dn,
  skipProcessing: pn,
  stripColorProfile: fn,
  stripMetadata: On,
  style: jn,
  trim: Pn,
  unsharpMasking: Mn,
  videoThumbnailSecond: Xt,
  watermark: An,
  watermarkShadow: Un,
  watermarkSize: Cn,
  watermarkText: Hn,
  watermarkUrl: Jn,
  webpOptions: Ln,
  width: es,
  wildOptions: ee,
  zoom: ns
}, Symbol.toStringTag, { value: "Module" })), T = {
  plain: !0,
  base64: !0,
  encrypted: !0
}, Js = (t, e) => {
  n(t.value, "url.value", "Must be a string"), n(
    t.type,
    "url.type",
    `Valid values are: ${Object.keys(T).join(", ")}`
  ), d(T, t.type, "url.type");
  let o = "";
  if (e)
    for (const [, s] of Object.entries(ls))
      s.test(e) && (o += "/", o += s.build(e));
  let r = "";
  return t.type === "plain" ? r = `/plain/${t.value}` : t.type === "base64" ? r = `/${t.value}` : t.type === "encrypted" && (r = `/enc/${t.value}`), `${o}${r}`;
}, oe = (t) => {
  if ("average" in t)
    return t.average;
  if ("avg" in t)
    return t.avg;
}, cs = (t) => oe(t) !== void 0, us = (t) => {
  const e = oe(t);
  n(e, "average"), n(e.average, "average.average");
  const o = e.ignore_transparent === void 0 ? "" : `:${l(e.ignore_transparent)}`;
  return `avg:${l(e.average)}${o}`;
}, ds = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: us,
  test: cs
}, Symbol.toStringTag, { value: "Module" })), re = (t) => {
  if ("blurhash" in t)
    return t.blurhash;
  if ("bh" in t)
    return t.bh;
}, bs = (t) => re(t) !== void 0, _s = (t) => {
  const e = re(t);
  n(e, "blurhash");
  const { x_components: o, y_components: r } = e;
  return n(o, "blurhash.x_components"), n(r, "blurhash.y_components"), i(o, "blurhash.x_components", {
    addParam: { min: 0, max: 9 }
  }), i(r, "blurhash.y_components", {
    addParam: { min: 0, max: 9 }
  }), `bh:${o}:${r}`;
}, ps = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: _s,
  test: bs
}, Symbol.toStringTag, { value: "Module" })), ne = (t) => {
  if ("detect_objects" in t)
    return t.detect_objects;
  if ("do" in t)
    return t.do;
}, ms = (t) => ne(t) !== void 0, gs = (t) => {
  const e = ne(t);
  return n(e, "detect_objects"), `do:${l(e)}`;
}, fs = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: gs,
  test: ms
}, Symbol.toStringTag, { value: "Module" })), se = (t) => {
  if ("dimensions" in t)
    return t.dimensions;
  if ("d" in t)
    return t.d;
}, $s = (t) => se(t) !== void 0, ys = (t) => {
  const e = se(t);
  return n(e, "dimensions"), `d:${l(e)}`;
}, Os = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: ys,
  test: $s
}, Symbol.toStringTag, { value: "Module" })), ie = (t) => {
  if ("dominant_colors" in t)
    return t.dominant_colors;
  if ("dc" in t)
    return t.dc;
}, vs = (t) => ie(t) !== void 0, hs = (t) => {
  const e = ie(t);
  n(e, "dominant_colors"), n(e.dominant_colors, "dominant_colors.dominant_colors");
  const o = e.build_missed === void 0 ? "" : `:${l(e.build_missed)}`;
  return `dc:${l(e.dominant_colors)}${o}`;
}, js = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: hs,
  test: vs
}, Symbol.toStringTag, { value: "Module" })), ae = (t) => {
  if ("exif" in t)
    return t.exif;
}, Ss = (t) => ae(t) !== void 0, ws = (t) => {
  const e = ae(t);
  return n(e, "EXIF"), `exif:${l(e)}`;
}, Ps = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: ws,
  test: Ss
}, Symbol.toStringTag, { value: "Module" })), le = (t) => {
  if ("format" in t)
    return t.format;
  if ("f" in t)
    return t.f;
}, zs = (t) => le(t) !== void 0, xs = (t) => {
  const e = le(t);
  return n(e, "format"), `f:${l(e)}`;
}, Ts = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: xs,
  test: zs
}, Symbol.toStringTag, { value: "Module" })), ce = (t) => {
  if ("iptc" in t)
    return t.iptc;
}, Ms = (t) => ce(t) !== void 0, ks = (t) => {
  const e = ce(t);
  return n(e, "IPTC"), `iptc:${l(e)}`;
}, Bs = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: ks,
  test: Ms
}, Symbol.toStringTag, { value: "Module" })), ue = (t) => {
  if ("palette" in t)
    return t.palette;
  if ("p" in t)
    return t.p;
}, qs = (t) => ue(t) !== void 0, Es = (t) => {
  const e = ue(t);
  if (n(e, "palette"), i(e, "palette", {
    addParam: { min: 0, max: 256, isInt: !0 }
  }), e === 1)
    throw new Error("palette option is should be 0 or between 2 and 256");
  return `p:${e}`;
}, Is = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Es,
  test: qs
}, Symbol.toStringTag, { value: "Module" })), de = (t) => {
  if ("size" in t)
    return t.size;
  if ("s" in t)
    return t.s;
}, As = (t) => de(t) !== void 0, Vs = (t) => {
  const e = de(t);
  return n(e, "size"), `s:${l(e)}`;
}, Fs = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Vs,
  test: As
}, Symbol.toStringTag, { value: "Module" })), be = (t) => {
  if ("video_meta" in t)
    return t.video_meta;
  if ("vm" in t)
    return t.vm;
}, Us = (t) => be(t) !== void 0, Ns = (t) => {
  const e = be(t);
  return n(e, "video_meta"), `vm:${l(e)}`;
}, Rs = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Ns,
  test: Us
}, Symbol.toStringTag, { value: "Module" })), _e = (t) => {
  if ("xmp" in t)
    return t.xmp;
}, Cs = (t) => _e(t) !== void 0, Qs = (t) => {
  const e = _e(t);
  return n(e, "XMP"), `xmp:${l(e)}`;
}, Ds = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  build: Qs,
  test: Cs
}, Symbol.toStringTag, { value: "Module" })), Hs = /* @__PURE__ */ Object.freeze(/* @__PURE__ */ Object.defineProperty({
  __proto__: null,
  average: ds,
  blurhash: ps,
  cacheBuster: N,
  crop: H,
  detectObjects: fs,
  dimensions: Os,
  dominantColors: js,
  exif: Ps,
  expires: Z,
  format: Ts,
  iptc: Bs,
  maxSrcFileSize: gt,
  maxSrcResolution: $t,
  page: jt,
  palette: Is,
  preset: xt,
  size: Fs,
  videoMeta: Rs,
  videoThumbnailSecond: Xt,
  wildOptions: ee,
  xmp: Ds
}, Symbol.toStringTag, { value: "Module" })), M = {
  plain: !0,
  base64: !0,
  encrypted: !0
}, Ys = (t, e) => {
  n(t.value, "url.value", "Must be a string"), n(
    t.type,
    "url.type",
    `Valid values are: ${Object.keys(M).join(", ")}`
  ), d(M, t.type, "url.type");
  let o = "";
  if (e)
    for (const [, s] of Object.entries(Hs))
      s.test(e) && (o += "/", o += s.build(e));
  let r = "";
  return t.type === "plain" ? r = `/plain/${t.value}` : t.type === "base64" ? r = `/${t.value}` : t.type === "encrypted" && (r = `/enc/${t.value}`), `${o}${r}`;
}, Gs = "/info";
export {
  Gs as INFO_PREFIX,
  Ys as generateImageInfoUrl,
  Js as generateUrl
};
